tls/certificate_store.go

```diff
@@ -31,3 +35,82 @@ func (c CertificateStore) GetAllDomains() []string {
 	}
 	return allCerts
 }
+
+// GetBestCertificate returns the best match certificate, and caches the response
+func (c CertificateStore) GetBestCertificate(domainToCheck string) *tls.Certificate {
+	if c.CertCache == nil {
+		c.CertCache = map[string]*tls.Certificate{}
+	}
+
+	if c.CertCache[domainToCheck] != nil {
+		return c.CertCache[domainToCheck]
+	}
+
+	matchedCerts := map[string]*tls.Certificate{}
+	if c.DynamicCerts != nil {
+		if c.DynamicCerts.Get() != nil {
+			for domains, cert := range c.DynamicCerts.Get().(map[string]*tls.Certificate) {
+				for _, certDomain := range strings.Split(domains, ",") {
+					if MatchDomain(domainToCheck, certDomain) {
+						matchedCerts[certDomain] = cert
+					}
+				}
+			}
+		}
+	}
+	if c.StaticCerts != nil {
+		if c.StaticCerts.Get() != nil {
+			for domains, cert := range c.StaticCerts.Get().(map[string]*tls.Certificate) {
+				for _, certDomain := range strings.Split(domains, ",") {
+					if MatchDomain(domainToCheck, certDomain) {
+						matchedCerts[certDomain] = cert
+					}
+				}
+			}
+		}
+	}
+	if len(matchedCerts) > 0 {
+		//sort map by keys
+		keys := make([]string, 0, len(matchedCerts))
+		for k := range matchedCerts {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		//cache best match
+		c.CertCache[domainToCheck] = matchedCerts[keys[len(keys)-1]]
```

Comment:
Cannot we check other information?

For example:
- Check if a certificate is self-signed and do not provide it if another signed by a recognized CA
- Check if a certificate is not still valid and do not provide it if another certifcate is still valid.

Reply:
We can, but did we want to include that in this PR? or did we want to add more checks etc in a separate PR?

Reply:
Because if we are inspecting the certs, we probably want to do a bit more checks such as the RSA/EC key checks etc...

Reply:
It is OK is we do that in a next PR :+1: